crux_macros/src/effect.rs

```diff
@@ -0,0 +1,98 @@
+use convert_case::{Case, Casing};
+use darling::{ast, util, FromDeriveInput, FromField, FromMeta};
+use proc_macro2::TokenStream;
+use quote::quote;
+use syn::{DeriveInput, Ident, Index, Type};
+
+#[derive(FromDeriveInput)]
+#[darling(attributes(effect), supports(struct_named))]
+struct EffectStructReceiver {
+    ident: Ident,
+    name: Option<Type>,
+    app: Option<Type>,
+    event: Option<Type>,
+    data: ast::Data<util::Ignored, EffectFieldReceiver>,
+}
+
+#[derive(Debug, FromField)]
+#[darling(attributes(effect))]
+pub struct EffectFieldReceiver {
+    ident: Option<Ident>,
+    operation: Option<Type>,
```

Comment:
re my "could we use an associated type for this" - you'd want to add a `ty: Type` field here and darling will use it to extract the type.

Then assuming the associated type was called `Operation` you can do a `quote! { <#ty as ::crux_core::Capability>::Operation }` to get at it.

Reply:
Oh OK! Now I get it. Very clever :-) 🤯